runtime/interface.cpp

```diff
@@ -1305,64 +1306,107 @@ static void parse_http_authorization_header(const string &header_value) {
   v$_SERVER.set_value(string("AUTH_TYPE"), auth_scheme);
 }
 
-static void init_superglobals(const char *uri, int uri_len, const char *get, int get_len, const char *headers, int headers_len, const char *post, int post_len,
-                              const char *request_method, int request_method_len, int remote_ip, int remote_port, int keep_alive,
-                              const int *serialized_data, int serialized_data_len, long long rpc_request_id, int rpc_remote_ip, int rpc_remote_port, int rpc_remote_pid, int rpc_remote_utime) {
-  rpc_parse(serialized_data, serialized_data_len);
+static void save_rpc_query_headers(const tl_query_header_t &header) {
+  namespace flag = vk::tl::common::rpc_invoke_req_extra_flags;
+
+  if (header.actor_id) {
+    v$_SERVER.set_value(string("RPC_ACTOR_ID"), static_cast<int64_t>(header.actor_id));
+  }
+  if (header.flags) {
+    v$_SERVER.set_value(string("RPC_EXTRA_FLAGS"), header.flags);
+  }
+  if (header.flags & flag::wait_binlog_pos) {
+    v$_SERVER.set_value(string("RPC_EXTRA_WAIT_BINLOG_POS"), static_cast<int64_t>(header.wait_binlog_pos));
+  }
+  if (header.flags & flag::string_forward_keys) {
+    array<string> string_forward_keys;
+    string_forward_keys.reserve(header.string_forward_keys.size(), 0, true);
+    for (const auto &str_key : header.string_forward_keys) {
+      string_forward_keys.emplace_back(string().assign(str_key.c_str()));
+    }
+    v$_SERVER.set_value(string("RPC_EXTRA_STRING_FORWARD_KEYS"), string_forward_keys);
+  }
+  if (header.flags & flag::int_forward_keys) {
+    array<int64_t> int_forward_keys;
+    int_forward_keys.reserve(header.int_forward_keys.size(), 0, true);
+    for (int int_key : header.int_forward_keys) {
+      int_forward_keys.emplace_back(int_key);
+    }
+    v$_SERVER.set_value(string("RPC_EXTRA_INT_FORWARD_KEYS"), int_forward_keys);
+  }
+  if (header.flags & flag::string_forward) {
+    v$_SERVER.set_value(string("RPC_EXTRA_STRING_FORWARD"), string().assign(header.string_forward.c_str()));
+  }
+  if (header.flags & flag::int_forward) {
+    v$_SERVER.set_value(string("RPC_EXTRA_INT_FORWARD"), static_cast<int64_t>(header.int_forward));
+  }
+  if (header.flags & flag::custom_timeout_ms) {
+    v$_SERVER.set_value(string("RPC_EXTRA_CUSTOM_TIMEOUT_MS"), header.custom_timeout);
+  }
+  if (header.flags & flag::supported_compression_version) {
+    v$_SERVER.set_value(string("RPC_EXTRA_SUPPORTED_COMPRESSION_VERSION"), header.supported_compression_version);
+  }
+  if (header.flags & flag::random_delay) {
+    v$_SERVER.set_value(string("RPC_EXTRA_RANDOM_DELAY"), header.random_delay);
+  }
+}
+
+static void init_superglobals(const http_query_data *http_data, const rpc_query_data *rpc_data) {
```

Comment:
nit: logically by code this is const references, not pointers, because we are sure, that they are always non nullptr